tests/components/droplet/test_config_flow.py

```diff
@@ -0,0 +1,221 @@
+"""Test Droplet config flow."""
+
+from ipaddress import IPv4Address
+from unittest.mock import MagicMock
+
+from homeassistant import config_entries
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResultType
+from homeassistant.helpers.service_info.zeroconf import ZeroconfServiceInfo
+
+from .conftest import MockClientBehaviors, mock_try_connect
+
+from tests.common import MockConfigEntry
+
+
+async def test_user_setup(
+    hass: HomeAssistant, mock_coordinator_setup: MagicMock
+) -> None:
+    """Test successful Droplet user setup."""
+    result = await hass.config_entries.flow.async_init(
+        "droplet", context={"source": config_entries.SOURCE_USER}
+    )
+    assert result is not None
+    assert result.get("type") is FlowResultType.FORM
+    assert result.get("step_id") == "user"
+
+    with mock_try_connect(MockClientBehaviors.GOOD):
+        result = await hass.config_entries.flow.async_configure(
+            result.get("flow_id"), user_input={"code": "123456", "host": "192.168.1.2"}
+        )
+        assert result is not None
+        assert result.get("type") is FlowResultType.CREATE_ENTRY
+        assert result.get("data") == {
+            "code": "123456",
+            "device_id": "Droplet-1234",
+            "host": "192.168.1.2",
+            "name": "Droplet",
+            "port": 443,
+        }
+
+
+async def test_user_setup_cannot_connect(hass: HomeAssistant) -> None:
+    """Test user setup when the device fails to connect at all."""
+    result = await hass.config_entries.flow.async_init(
+        "droplet", context={"source": config_entries.SOURCE_USER}
+    )
+    assert result is not None
+    assert result.get("type") is FlowResultType.FORM
+    assert result.get("step_id") == "user"
+
+    with mock_try_connect(MockClientBehaviors.FAIL_OPEN):
+        result = await hass.config_entries.flow.async_configure(
+            result.get("flow_id"), user_input={"code": "123456", "host": "192.168.1.2"}
+        )
+        assert result is not None
+        assert result.get("type") is FlowResultType.FORM
+        assert result.get("errors") == {"base": "cannot_connect"}
+
+
+async def test_user_setup_no_device_id(hass: HomeAssistant) -> None:
+    """Test user setup when the device connects but fails to send its ID."""
+    result = await hass.config_entries.flow.async_init(
+        "droplet", context={"source": config_entries.SOURCE_USER}
+    )
+    assert result is not None
+    assert result.get("type") is FlowResultType.FORM
+    assert result.get("step_id") == "user"
+
+    with mock_try_connect(MockClientBehaviors.NO_DEVICE_ID):
+        result = await hass.config_entries.flow.async_configure(
+            result.get("flow_id"), user_input={"code": "123456", "host": "192.168.1.2"}
+        )
+        assert result is not None
+        assert result.get("type") is FlowResultType.FORM
+        assert result.get("errors") == {"base": "cannot_connect"}
```

Comment:
These flows can be parametrized, also, they should recover and end in a CREATE_ENTRY to show that this flow is able to recover